### DISCUSSION TITLE
How to find where the material derivative is defined

### DISCUSSION THREAD

**Discussion by YXY101676** [idaholab/moose#29553] [Q&A Modules: Phase field]
Dears
In the following code, many material derivatives such as _dLdop and _dLdarg[i] are called. I cannot find definition of declarePropertyDerivative in GBEvolution or  GBEvolutionbase.
```
registerMooseObject("PhaseFieldApp", ACInterface);
InputParameters
ACInterface::validParams()
{
  InputParameters params = JvarMapKernelInterface<Kernel>::validParams();
  params.addClassDescription("Gradient energy Allen-Cahn Kernel");
  params.addParam<MaterialPropertyName>("mob_name", "L", "The mobility used with the kernel");
  params.addParam<MaterialPropertyName>("kappa_name", "kappa_op", "The kappa used with the kernel");
  params.addParam<bool>("variable_L",
                        true,
                        "The mobility is a function of any MOOSE variable (if "
                        "this is set to false L must be constant over the "
                        "entire domain!)");
  return params;
}

ACInterface::ACInterface(const InputParameters & parameters)
  : DerivativeMaterialInterface<JvarMapKernelInterface<Kernel>>(parameters),
    _L(getMaterialProperty<Real>("mob_name")),
    _kappa(getMaterialProperty<Real>("kappa_name")),
    _variable_L(getParam<bool>("variable_L")),
    _dLdop(getMaterialPropertyDerivative<Real>("mob_name", _var.name())),
    _d2Ldop2(getMaterialPropertyDerivative<Real>("mob_name", _var.name(), _var.name())),
    _dkappadop(getMaterialPropertyDerivative<Real>("kappa_name", _var.name())),

_dLdarg[i] = &getMaterialPropertyDerivative<Real>("mob_name", i);
```
GBEvolution  is the main computational item in materials.

**Comment by GiudGiud** [comment]
Hello

The `declarePropertyDerivative` are part of the DerivativeMaterialInterface base class.
It is a template for making these properties, that we instantiate here with the `JvarMapKernelInterface<Kernel>` class (itself a template instantiation)

Guillaume

**Reply by YXY101676** [reply]
I know that the DerivativeMaterialInterface base class is the source of declarePropertyDerivative, but what I want to know is where _dLdop is instantiated in which file.  Similar to
 _D_elastic_tensor[op_index] = &declarePropertyDerivative<RankFourTensor>(_elasticity_tensor_name, coupledName("v", op_index));
And I can't find the JvarMapKernelInterface<Kernel> class file.

**Reply by GiudGiud** [reply]
> And I can't find the JvarMapKernelInterface class file.

it's in a header as well, JvarMapInterface.h

> where _dLdop is instantiated in which file

you can use grep to find that. See ACBulk.h or ACInterface.C

**Reply by YXY101676** [reply]
I can only find the relevant calls for _dLdop(getMaterialPropertyDerivative<T>("mob_name", _var())) in ACBulk.h, but I cannot find the relevant content for _dLdop(declareMaterialPropertyDerivative<T>("mob", _var.name())).The same situation occurs in ACInterface.C.

**Reply by YXY101676** [reply]
All related material calculations are in GBEvolution.C, and I couldn't find any.

**Reply by GiudGiud** [reply]
if you find a simulation in the repository that uses this property, use Debug/show_material_props=true to see which Material declares it

**Reply by YXY101676** [reply]
![image](https://github.com/user-attachments/assets/e105812a-57d1-4893-8161-8be948e85429)

![image](https://github.com/user-attachments/assets/ec140da7-8976-4db8-8b03-6b2e83740e57)
The kernel ACinterface exists, but the material property _dLdop is not displayed. What's the situation?

**Reply by GiudGiud** [reply]
dLdop is most likely created by the material creating L
